docker/DOCKER.md

```diff
@@ -0,0 +1,214 @@
+# Docker Images for the balena CLI
+
+Docker images with the balena CLI and docker-in-docker.
+
+## Features Overview
+
+These CLI images are based on the popular [Balena base images](https://www.balena.io/docs/reference/base-images/base-images/)
+so they include many of the features you see there.
+
+- Multiple Architectures:
+    - `rpi`
+    - `armv7hf`
+    - `aarch64` (debian only)
+    - `amd64`
+    - `i386`
+- Multiple Distributions
+    - `debian`
+    - `alpine`
+- [cross-build](https://www.balena.io/docs/reference/base-images/base-images/#building-arm-containers-on-x86-machines) functionality for building ARM containers on x86.
+- Helpful package installer script called `install_packages` inspired by [minideb](https://github.com/bitnami/minideb#why-use-minideb).
+
+Note that there are some additional considerations when running the CLI via Docker so pay close attention to the [Basic usage](#basic-usage) and
+[Advanced Usage](#advanced-usage) sections for examples of different CLI commands.
+
+## Image Names
+
+`balenalib/<arch>-<distro>-balenacli:<cli_ver>`
+
+- `<arch>` is the architecture and is mandatory. If using `Dockerfile.template`, you can replace this with `%%BALENA_ARCH%%`.
+For a list of available device names and architectures, see the [Device types](https://www.balena.io/docs/reference/base-images/devicetypes/).
+- `<distro>` is the Linux distribution and is mandatory. Currently there are 2 distributions, namely `debian` and `alpine`.
+
+## Image Tags
+
+In the tags, all of the fields are optional, and if they are left out, they will default to their `latest` pointer.
+
+- `<cli_ver>` is the version of the balena CLI, for example, `12.40.2`, it can also be substituted for `latest`.
+
+To ensure that the latest available CLI image is used, it is also necessary
+to add the `--pull` flag to `docker run` commands. Otherwise, Docker will
+reuse a locally cached image, without checking for available updates.
+
+## Examples
+
+`balenalib/amd64-debian-balenacli:12.40.2`
+
+- `<arch>`: amd64 - suitable for running on most workstations
+- `<distro>`: debian - widely used base distro
+- `<cli_ver>`: 12.40.2
+
+`balenalib/armv7hf-alpine-balenacli`
+
+- `<arch>`: armv7hf - suitable for running on a Raspberry Pi 3 for example
+- `<distro>`: alpine - smaller footprint than debian
+- `<cli_ver>`: omitted - the latest available CLI version will be used
+
+## Basic Usage
+
+Here's a small example of running a single, detached container
+in the background and using `docker exec` to run balena CLI commands.
```

Comment:
NIT, to help novice users understand it the interplay between `docker run` and `docker exec` when running the CLI in this manner.

-------------
\## Basic Usage

The recommended basic usage consists of running a single, detached container
named "cli" in the background, with the `docker run` command:

```
$ docker run --name cli --detach --pull always --privileged --network host --rm -it balenalib/amd64-debian-balenacli /bin/bash
...
Status: Image is up to date for balenalib/amd64-debian-balenacli
3fd0fd2c86049c5c6f6dad665aef0e6a0ffc83003d73a3522924f844270dbde4

$ docker ps -a
CONTAINER ID   IMAGE                              NAMES
3fd0fd2c8604   balenalib/amd64-debian-balenacli   cli

$ docker logs cli
...
INFO[2021-03-10T21:14:48.826556111Z] Daemon has completed initialization
INFO[2021-03-10T21:14:48.838769027Z] API listen on /var/run/docker.sock
```

Then, use `docker exec` to run balena CLI commands in that container, as follows.

```
< docker exec command examples go here >
```

When the background container is no longer required, it can be stopped with:
```
$ docker stop cli
cli
```
-------------

Reply:
Then again, I am having second thoughts about the usefulness of this "basic usage" with `docker run` + `docker exec`, mainly because we are not giving examples of how to share the filesystem between "host OS" and the CLI containers, and for sure users will be trying to run commands like `balena push`, especially if they take "basic usage" to mean "essential usage" (`balena push` is one of the most essential commands).

We would also have to add a note about the fact that `balena login` web authentication (that the CLI highlights as the recommended option) won't work at all. Which is OK given the container limitations, but a note would be needed.

How bad would it be to bind mount the user's homedir to `/root` in the container? Just tested and got `nvm` errors (script running from `.profile` / `.bash_profile` / `.bashrc` files). Probably not the best idea, but perhaps we could document bind mounting a specific folder for data sharing between host OS and container.

I seem to remember that `balena preload` does not work if the source image is loaded through a bind mount (it had to be copied to the container's internal filesystem). If so, we would have to document that too, otherwise users will be wasting time attempting and will bring the matter to support...

Reply:
> We would also have to add a note about the fact that balena login web authentication (that the CLI highlights as the recommended option) won't work at all

Good point, this will never work even by copying the URL as the callback auth hook won't be able to reach the CLI. I'll add a note.

> How bad would it be to bind mount the user's homedir to /root in the container?

I wouldn't suggest doing this, since the container runs as root and your host OS user would end up losing access to some important files. We can show an example of a named volume to persist cache and credentials if it helps?

> I seem to remember that balena preload does not work if the source image is located in a bind mount (it had to be copied to the container's internal filesystem).

Preload does work in a bind mount, and I have an example farther down, but the absolute path to the image file must be the same from the perspective of the docker host and the CLI. So it only matters if using the host docker socket, and can be managed by mounting the path to the image on the host to an identical path in the container (eg. `-v $PWD:$PWD -w $PWD`).

Reply:
Stepping back and thinking from the point of view of the end user... As an end user, I would like the DOCKER.md guide to show me how I am supposed to use the CLI's docker images for, say, at least the 15 "primary commands" listed with `balena help` (as opposed to the "additional commands" listed with `balena help -v`) -

```
$ balena help
...
PRIMARY COMMANDS
  login                                  login to balena
  push <applicationOrDevice>             start a build on the remote balenaCloud build servers, or a local mode device
  logs <device>                          show device logs
  ssh <applicationOrDevice> [service]    SSH into the host or application container of a device
  apps                                   list all applications
  app <nameOrSlug>                       display information about a single application
  devices                                list all devices
  device <uuid>                          show info about a single device
  tunnel <deviceOrApplication>           tunnel local ports to your balenaOS device
  preload <image>                        preload an app on a disk image (or Edison zip archive)
  build [source]                         build a project locally
  deploy <appName> [image]               deploy a single image or a multicontainer project to a balena application
  join [deviceIpOrHostname]              move a local device to an application on another balena server
  leave [deviceIpOrHostname]             remove a local device from its balena application
  scan                                   scan for balenaOS devices on your local network
```

It doesn't matter so much if we are using `docker run` + `docker exec`, or just `docker run`. I think that using `docker exec` was supposed to be a performance optimization (so that the docker daemon doesn't get started again and again) and a way of persisting the `login` information across multiple CLI commands. This is good, as long as the 15 primary commands can be used this way, and we have examples.

DOCKER.md is currently split between Basic Usage and Advanced Usage. I think it doesn't matter if we have one or two such sections and which section uses `docker exec`. The sections could be swapped, e.g. Basic Usage could use `docker run` only, and Advanced Usage could use `docker exec`, or one of the 2 sections could be deleted or shortened if they are redundant (meaning: no benefit over the alternative usage). Whichever section is titled Basic Usage, I think it should cover all 15 primary commands, with some "usability requirements" if possible:

- We should document that users have the option of using a host Docker Engine or a Docker Engine inside the CLI container, and how to switch between them. A specific note about the `deploy` command would be nice, as it deploys a pre-built image through the chosen Docker Engine.
- If possible, the project data for push/build/deploy should be outside the container, which implies a bind mount. An example should be provided of bind mounting.
- We should document the restrictions on `balena login` (that web auth doesn't work).
- We should document your explanation re preload that makes a lot of sense:
> Preload does work in a bind mount, but the absolute path to the image file must be the same from the perspective of the docker host and the CLI. So it only matters if using the host docker socket, and can be managed by mounting the path to the image on the host to an identical path in the container (eg. -v $PWD:$PWD -w $PWD).

Much of the above is already done / available, it's just a matter of moving things around so it's easier for users.

Reply:
My latest push just added some much better notes to each of the example commands, but I'm definitely missing some of the top 15 like `push`, `logs, `apps`, `devices`, `device`, `tunnel` (this one is tricky and requires some explanation), `join`, `leave`.

Reply:
Some of them are trivial because it's just API queries -- apps, devices, device... Others have forms that take the IP address of a local device and will also probably work without problems: logs, join, leave. The trickier ones that require bind mounting are push/build and `deploy --build`. The `deploy` command without the `--build` flag will load a pre-build image using some Docker Engine. The `tunnel` command may not work at all and that's fine, as long as we add a note so that users won't spend time trying it and then come to support and say "it's broken!"

Reply:
Sounds good! Avoiding duplication is good! 😁 
All the "purely API" commands could be combined as there's nothing special about them, like bind mounting or docker sockets... `apps | app | devices | device`

Reply:
I'm dropping the current exec "basic" usage example as the container definition needs to change for many commands. Instead, I'm providing two examples for each command, one with a named volume for credentials and one with a shell session and multiple commands.

eg.
### login

- <https://www.balena.io/docs/reference/balena-cli/#login>

The `balena login` command cannot be used with web authorization and a browser
when running in a container. Instead it must be used with `--token` or `--credentials`.

Notice that in this example we've used a named volume `balena_cache` to store credentials
for future runs of the image. This is optional but avoids having to run the login
command again for each of the other commands.

```bash
$ docker volume create balena_cache

# example 1: login with credentials
$ docker run --pull always --rm -it \
    -v balena_cache:/root/.balena \
    balenalib/amd64-debian-balenacli login --credentials --email johndoe@gmail.com --password secret

# example 2: login with token
$ docker run --pull always --rm -it \
    -v balena_cache:/root/.balena \
    balenalib/amd64-debian-balenacli login --token "..."
```

### app | apps

- <https://www.balena.io/docs/reference/balena-cli/#app-nameorslug>
- <https://www.balena.io/docs/reference/balena-cli/#apps>

```bash
# example 1: use existing credentials from login example
$ docker run --pull always --rm -it \
    -v balena_cache:/root/.balena \
    balenalib/amd64-debian-balenacli app myApp

# example 2: start a shell session and execute multiple commands
$ docker run --pull always --rm -it \
    balenalib/amd64-debian-balenacli bash

> balena login --credentials --email johndoe@gmail.com --password secret
> balena app myApp
> exit
```

Reply:
It's looking good! 👍 
What do you think about running `dockerd` conditionally on an env var that was defined through  
`docker run --env DOCKERD=1`  
instead of checking for a bind mount of `docker.sock`? The reasoning being that most (?) commands don't use docker, like `login` and `apps`, and yet the Docker Engine starts verbosely every time. It would make some `docker run` command lines even longer than they already are, but any user who regularly wanted to use the CLI this way would for sure write a 2-line script or `alias` shell command to save typing.

Then for commands that do use docker like `build` and `preload`, the "2 examples" could include an example with a bind mount of `docker.sock` and no env var, and another example with `--env DOCKERD=1`.
Just a thought!

Reply:
It's a good thought, I'm on it!

Reply:
Awesome job with the latest DOCKER.md documentation, Kyle! 💯  🥇 
It's looking great! 👌

Reply:
Thank you! :blush: